### PR TITLE
add how it works section

### LINKED ISSUE #182: Add "how it works" page
There should be a page explaining the basic idea, i.e. that an expression is a function from a dataframe to a Series

For example, if we start with `pd.DataFrame({'a': [1,2,3], 'b': [4,5,6]})`, then `nw.col('a')` takes `df` as input and returns `df['a']`

We can access that function with `._col`

e.g.:
```python
In [1]: import pandas as pd

In [2]: from narwhals._pandas_like.dataframe import PandasDataFrame

In [3]: from narwhals._pandas_like.namespace import PandasNamespace

In [4]: df = PandasDataFrame(pd.DataFrame({'a': [1,2,3], 'b': [4,5,6]}), implementation='pandas')

In [5]: pn = PandasNamespace(implementation='pandas')

In [6]: pn.col('a')._call(df)[0]._series
Out[6]:
0    1
1    2
2    3
Name: a, dtype: int64
```

Or:
```python
In [7]: (pn.col('a') + 1)._call(df)[0]._series
Out[7]:
0    2
1    3
2    4
Name: a, dtype: int64
```

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #182

### CODE REVIEW

File: narwhals/dependencies.py [hunk 1]
```diff
@@ -18,8 +18,7 @@ def get_pandas() -> Any:
 
 def get_modin() -> Any:  # pragma: no cover
     """Import modin (if available - else return None)."""
-    modin = sys.modules.get("modin", None)
-    if modin is not None:
+    if (modin := sys.modules.get("modin", None)) is not None:
```
Comment (MarcoGorelli) [inline narwhals/dependencies.py:21]: drive-by walrus

File: docs/how_it_works.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+from narwhals._pandas_like.namespace import PandasNamespace
+
+pn = PandasNamespace(implementation='pandas')
+nw.col('a')._call(pn)
+```
+The result from the last line above is the same as we'd get from `pn.col('a')`, and it's
+a `narwhals._pandas_like.expression.PandasExpr` object, which we'll call `PandasExpr` for
+short.
+
+`PandasExpr` also have a `_call` method - but this one expects a `PandasDataFrame` as input.
```
Comment (FBruzzesi) [inline docs/how_it_works.md:97]: Nit:

```suggestion
`PandasExpr` also has a `_call` method - but this one expects a `PandasDataFrame` as input.
```

File: docs/how_it_works.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+Recall from above that an expression is a function from a dataframe to a sequence of series.
+The `_call` method gives us that function! Let's see it in action.
+
+Note: the following examples uses `PandasDataFrame` and `PandasSeries`. These are wrappers
+around pandas DataFrame and pandas Series, which are Narwhals-compliant. To get the native
+pandas objects out from inside them, we access `PandasDataFrame._dataframe` and `PandasSeries._series`.
```
Comment (FBruzzesi) [inline docs/how_it_works.md:103]: Not the biggest fan of this phrasing. Maybe something like:

```suggestion
around pandas DataFrame and pandas Series, which are Narwhals-compliant. We can access the 
pandas objects via `PandasDataFrame._dataframe` and `PandasSeries._series`.
```

File: docs/how_it_works.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+import narwhals as nw
+from narwhals._pandas_like.namespace import PandasNamespace
+from narwhals._pandas_like.dataframe import PandasDataFrame
+import pandas as pd
+
+pn = PandasNamespace(implementation='pandas')
+
+df_pd = pd.DataFrame({'a': [1,2,3], 'b': [4,5,6]})
+df = PandasDataFrame(df_pd, implementation='pandas')
+expression = pn.col('a') + 1
```
Comment (FBruzzesi) [inline docs/how_it_works.md:115]: Should we use some fancy mkdocs-material to highlight the fact that this such line does not compute anything as is? Maybe [code annotation](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#code-annotations)

File: docs/how_it_works.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+0    2
+1    3
+2    4
+Name: a, dtype: int64
+```
+
+So indeed, our expression did what it said on the tin - it took some dataframe, took
+column 'a', and added 1 to it.
+
+If you search for `def register_expression_call`, you'll see that that's all
```
Comment (FBruzzesi) [inline docs/how_it_works.md:130]: Let's handle the link ourselves:

```suggestion
If you search for [`register_expression_call`](https://github.com/narwhals-dev/narwhals/blob/f18c533a0c2ff870ad489a6f93bca8bef2a98eda/narwhals/_pandas_like/utils.py#L171){target="_blank"}, you'll see that that's all
```

**Review by FBruzzesi — COMMENTED** [review]
That's a great write up! I got a much better understanding of the situation and internals 

The only addition I would do it to mention somewhere that pandas implementation is used also for the pandas-like APIs.

Other comments are just food for thought :)


ThanksThankYouGIF

File: docs/how_it_works.md (hunk 5 above)
Comment (MarcoGorelli) [reply docs/how_it_works.md:131]: the `{target="_black"}` didn't work for me when I rendered this. in any case, are sure? I think the link would quickly go out-of-date

**Comment by MarcoGorelli** [comment]
thanks Francesco for taking a look! I've addressed most comments, those I wasn't able to do shall we leave them for a follow-up?

File: docs/how_it_works.md (hunk 5 above)
Comment (FBruzzesi) [reply docs/how_it_works.md:131]: The one I added is the permalink, not just the link to main and line, which in theory should be more stable to changes in the file. `{target="_blank"}` should open on a new tab/page, which is something I prefer but can live without 😂

Comment (MarcoGorelli) [reply docs/how_it_works.md:131]: do i need to enable some extra feature for it to work? if i copy and paste your snippet, i get:

image

**Comment by MarcoGorelli** [comment]
going to merge this as I wanted to link this to somebody, we can then address the permalink (https://github.com/narwhals-dev/narwhals/pull/188#discussion_r1615621451) in a separate PR

thanks for your review!

File: docs/how_it_works.md (hunk 5 above)
Comment (FBruzzesi) [reply docs/how_it_works.md:131]: I will take a look, but if you want to merge it, let's remove it for now 😇
